Add MVP documentation (#42147)
kubernetes/website: content/en/docs/concepts/architecture/unknown-version-interoperability-proxy.md

---
reviewers:
- jpbetz
title: Unknown Version Interoperability Proxy
content_type: concept
weight: 220
---

<!-- overview -->
{{< feature-state state="alpha"  for_k8s_version="v1.28" >}}

Kubernetes {{< skew currentVersion >}} includes a feature that lets one
{{< glossary_tooltip text="kube-apiserver" term_id="kube-apiserver" >}}
proxy a resource request to a different _peer_ API server. This is useful when there are multiple
API servers running different versions of Kubernetes in one cluster (for example, during a long-lived
rollout to a new release of Kubernetes).

> This enables cluster administrators to configure highly available clusters that can be upgraded more safely by directing resource requests(made during the upgrade) to the correct kube-apiserver thus preventing incorrect 404("Not Found") errors.  

Reviewer: (nit)
Try:
```suggestion
This enables cluster administrators to configure highly available clusters that can be upgraded
more safely, by directing resource requests (made during the upgrade) to the correct kube-apiserver.
That proxying prevents users from seeing unexpected 404 Not Found errors that stem
from the upgrade process.
```

## Enabling the Unknown Version Interoperability Proxy
Ensure that `UnknownVersionInteroperabilityProxy` [feature gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) 
is enabled when you start the {{< glossary_tooltip text="API Server" term_id="kube-apiserver" >}}:

```shell
kube-apiserver \
--feature-gates=UnknownVersionInteroperabilityProxy=true \

> # required flags for this feature

Reviewer: (nit)
```suggestion
# required command line arguments for this feature
```
Outside of the Go ecosystem, “flag” refers specifically to command line arguments that control booleans (or sometimes tristates)

* To authenticate the destination server's serving certs, you must configure a certificate authority bundle by specifying the `--peer-ca-file` command line argument to the **source** API server.

### Configuration for peer API server connectivity

To find the network location of the peer kube-apiserver, you use the 
`--peer-advertise-ip` and `--peer-advertise-port` command line arguments to kube-apiserver or specify 
these fields in the API server configuration file.
If these flags are unspecified, the kube-apiserver uses the value from either `--advertise-address` or
`--bind-address` flag. If those too, are unset, the host's default interface is used.

## Interoperability proxying

> When you enable interoperability proxying, the [aggregation layer](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/) 

Reviewer: ```suggestion
When you enable interoperability proxying, the [aggregation layer](/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/) 
```

loads a special filter that does the following:

* When a resource request reaches an API server that cannot serve that API (either because it is at a version pre-dating the introduction of the API or the API is turned off on the API server) the API server attempts to send the request to a peer API server that can serve the requested API. It does so by identifying API groups / versions / resources that the local server doesn't recognise, and tries to proxy those requests to a peer API server that is capable of handling the request.
* If the peer API server fails to respond, the _source_ API server responds with 503("Service Unavailable") error.

### How it works under the hood

> When a kube-apiserver receives a resource request, it first checks which API servers can serve the requested resource. This check happens using the internal [`StorageVersion` API](https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/apis/apiserverinternal/types.go#L26-L37).

Reviewer: Can we either:
- link to the v1.28 release and make the link go to the right line
- omit the link and the internals details

?
Usually, if you can make a change to a detail and still pass conformance (or some imagined future conformance check on the GA equivalent), we don't need to mention it in docs.

Author: Oh I was trying to provide some reference to StorageVersion API as you had suggested earlier [here](https://github.com/kubernetes/website/pull/41723#discussion_r1271624569) . Also v1.28 release branch is not yet cut, so I can't link to the v1.28 release

Author: Omitted it for now.

* If the resource is known to the local API server (ex: `GET /api/v1/pods/some-pod`), the request is handled locally. 

* If there is no internal `StorageVersion` object found for the requested resource (ex: `GET /my-api/v1/my-resource`) and the configured APIService specifies proxying to an extension API server, that proxying happens following the usual

> [flow](https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/) for

Reviewer: ```suggestion
[flow](/docs/tasks/extend-kubernetes/configure-aggregation-layer/) for
```

extension APIs.

* If a valid internal `StorageVersion` object is found for the requested resource (ex: `GET /batch/v1/jobs`) and the API server trying to handle the request (the _handling API server_) has the `batch` API disabled, then the _handling API server_fetches the peer API servers that do serve the relevant API group / version / resource (`api/v1/batch` in this case) using the information in the fetched `StorageVersion` object. The _handling API server_ then proxies the request to one of the matching peer kube-apiservers that are aware of the requested resource.
  * If there is no peer known for that API group / version / resource, the handling API server passes the request to its own handler chain which should eventually return a 404("Not Found") response.

>   * If the handling API server has identified and selected a peer API server, but that peer fails to respond (for reasones such as network connectivity, or a data race between the request being received and a controller registering the peer's info into an internal `ServerStorageVersion` object), then the handling API server responds with a 503 (“Service Unavailable”) error.

Reviewer: Try this:
```suggestion
  * If the handling API server has identified and selected a peer API server, but that peer fails
    to respond (for reasons such as network connectivity issues, or a data race between the request
    being received and a controller registering the peer's info into the control plane), then the handling
    API server responds with a 503 (“Service Unavailable”) error.
```
Wrapping is helpful, and “reasones” is a typo.

---
reviewers:
- jpbetz
title: Unknown Version Interoperability Proxy
content_type: concept
weight: 220
---

<!-- overview -->
{{< feature-state state="alpha"  for_k8s_version="v1.28" >}}

Kubernetes {{< skew currentVersion >}} includes an alpha feature that lets one
{{< glossary_tooltip text="kube-apiserver" term_id="kube-apiserver" >}}
proxy a resource request to a different _peer_ API server. This is useful when there are multiple
API servers running different versions of Kubernetes in one cluster (for example, during a long-lived

> rollout to a new release of Kubernetes).

Reviewer 2: Remove "alpha here" since we already list the stability level above? (and avoid having one more thing to remember to update when we graduate this feature).

Reviewer 2: This covers "what" and "when" Can we say more about "who", "how" and "why"? E.g. "this allows cluster administrators to configure highly available clusters to ... during upgrade by ... "?

Author: > Remove "alpha here" since we already list the stability level above? (and avoid having one more thing to remember to update when we graduate this feature).
Done

> Can we say more about "who", "how" and "why"?
Added

Reviewer: > Remove "alpha here" since we already list the stability level above? (and avoid having one more thing to remember to update when we graduate this feature).

I suggested that text, and it's intentional. When we update to stable we'd omit the `{{< skew currentVersion >}}` shortcode. Someone looking at different docs versions will see “Kubernetes 1.28 ... alpha”, “Kubernetes 1.29 ... alpha”, “Kubernetes 1.30 ... beta” and finally ”Kubernetes lets you configure proxying&hellip;” once it's stable.

There's another reason. In code, it's good not to repeat ourselves. In tech docs, it can help to say things twice.

Author: sure, reverted back to original.

> This mechanism is called the _Unknown Version Interoperability Proxy_.

Reviewer 2: Remove? This line feels redundant.

Author: Done

Reviewer: Hi. Docs tech lead here.
Please put that text back. It will help readers to understand how that text links to the page, and wraps up the intro.

@Richabanker sorry for the mixed messaging on this.

Author: No worries, updated.

## Enabling the Unknown Version Interoperability Proxy
The Unknown Version Interoperability Proxy feature is controlled by a
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
and is not enabled by default. You can activate Unknown Version Interoperability Proxy by enabling

> the feature gate named `UnknownVersionInteroperabilityProxy` when you start the

Reviewer 2: Be concise here? Maybe just "Ensure the `UnknownVersionInteroperabilityProxy` [feature gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) is enabled" ?

Author: Done

{{< glossary_tooltip text="API Server" term_id="kube-apiserver" >}}:

```shell
kube-apiserver \
--feature-gates=UnknownVersionInteroperabilityProxy=true \
# required flag for this feature
--peer-ca-file=`ca bundle to verify serving certs of kube-apiserver`
# optional flags for this feature
--peer-advertise-ip=`ip of this kube-apiserver that should be used by peers to proxy requests`

> --peer-advertise-port=`port of this kube-apiserver that should be used by peers to proxy requests`

Reviewer 2: Would it be more helpful to readers to show the flags with valid inputs?

Should we be linking to/from:

- https://kubernetes.io/docs/tasks/administer-cluster/certificates/
- https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology/
- https://kubernetes.io/docs/tasks/administer-cluster/cluster-upgrade/ 

or similar?

Author: Did you mean specifying an example path for the cert files as inputs?

I added a link to [proxy cert setup](https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/#kubernetes-apiserver-client-authentication) for users to refer to better understand the --proxy-* and the --requestheader-client-ca flags

Author: Linking *from* https://kubernetes.io/docs/tasks/administer-cluster/cluster-upgrade/ makes sense to me. Not sure about the other two..

# …and other flags as usual
```

> ## Interoperability proxying

Reviewer 2: A picture would be worth 1000 words for sure here. Can we add a simplified diagram that highlights how a proxied request works (from a cluster administrator's perspective, where the apiserver binary is a black box)?

Author: ![image](https://github.com/kubernetes/website/assets/26771552/07857533-bf53-4a54-8545-cd8eb8c8b41a)

Author: How does this look?

Reviewer: (not needed for alpha, great to have though)

Reviewer: For tips, see https://www.k8s.dev/blog/2021/12/01/improve-your-documentation-with-mermaid.js-diagrams/

Author: ![image](https://github.com/kubernetes/website/assets/26771552/67802727-ac6b-4863-8d66-56203e92adde)

tried my hand at mermaid

When you enable interoperability proxying, the [aggregation layer](/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/) 
loads a special filter that identifies API groups / versions / resources that the local server doesn't recognise,

> and tries to proxy those requests to a peer API server that is capable of handling the request.

Reviewer 2: nit: Most users don't care (and shouldn't need to care) how the proxying happens in the guts of the apiserver. Consider instead focusing on what happens to the request?  "When a request reaches an API server that cannot serve that API (either because it is at a version pre-dating the introduction of the API or the API is turned off...) the API server attempts to send the request to a peer...." ?

Author: I added a small description on what happens to the request, and another "how it works under the hood" section highlighting more details on how the feature works.

specifies proxying to an extension API server, that proxying happens following the usual
[flow](https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/) for
extension APIs.

Once a valid internal `StorageVersion` object is found for the requested resource, 
the API server trying to handle the request (the _handling API server_) fetches the 
peer API servers that do serve the relevant API group / version / resources using the information 
in the fetched `StorageVersion` object. The _handling API server_ then proxies the request
to one of the matching peer kube-apiservers that are aware of the requested resource.
If there is no peer known for that API group / version / resource, the handling API server passes the 
request to its own handler chain which should eventually return a 404("Not Found") response.
If the handling API server has identified and selected a peer API server, but that peer fails to respond
(for reasones such as network connectivity, or a data race between the request being received and
a controller registering the peer's info into an internal `ServerStorageVersion` object), then the

> handling API server responds with a 503 (“Service Unavailable”) error.

Reviewer 2: Thread an example into this? It might be hard for someone unfamiliar with the idea to unpack all of this.

Author: Added different examples for different scenarios PTAL

## Proxy transport and authentication between API servers {#transport-and-authn}

To authenticate the destination server's serving certs, you must configure a certificate authority
bundle by specifying the `--peer-ca-file` command line argument to the **source** API server.

The source kube-apiserver reuses the existing `--proxy-client-cert-file` and `--proxy-client-key-file` to
present its identity that will be verified by its peer (the destination kube-apiserver). The destination API
server verifies that peer connection based on the configuration you specify using the
`--requestheader-client-ca-file` command line argument.

## Configuration for peer API server connectivity

To find the network location of the peer kube-apiserver, you use the 
`--peer-advertise-ip` and `--peer-advertise-port` command line arguments to kube-apiserver or specify 
these fields in the API server configuration file.
If these flags are unspecified, the kube-apiserver uses the value from either `--advertise-address` or

> `--bind-address` flag. If those too, are unset, the host's default interface is used.

Reviewer 2: Combine with the above `Enabling the Unknown Version Interoperability Proxy` section so that all the setup/configuration is explained together?

Author: Done
